test/components/connect.spec.js

```diff
@@ -1531,35 +1491,82 @@ describe('React', () => {
       expect(decorated.foo).toBe('bar')
     })
 
-    it('should use the store from the props instead of from the context if present', () => {
+    it('should use a custom context provider and consumer if given as an option to connect', () => {
       class Container extends Component {
         render() {
           return <Passthrough />
         }
       }
 
+      const context = React.createContext(null)
+
       let actualState
 
       const expectedState = { foos: {} }
+      const ignoredState = { bars: {} }
+
+      const decorator = connect(
+        state => {
+          actualState = state
+          return {}
+        },
+        undefined,
+        undefined,
+        { context }
+      )
+      const Decorated = decorator(Container)
+
+      const store1 = createStore(() => expectedState)
+      const store2 = createStore(() => ignoredState)
+
+      rtl.render(
+        <ProviderMock context={context} store={store1}>
+          <ProviderMock store={store2}>
+            <Decorated />
+          </ProviderMock>
+        </ProviderMock>
+      )
+
+      expect(actualState).toEqual(expectedState)
+    })
+
+    it('should use a custom context provider and consumer if passed as a prop to the component', () => {
+      class Container extends Component {
+        render() {
+          return <Passthrough />
+        }
+      }
+
+      const context = React.createContext(null)
+
+      let actualState
+
+      const expectedState = { foos: {} }
+      const ignoredState = { bars: {} }
+
       const decorator = connect(state => {
         actualState = state
         return {}
       })
       const Decorated = decorator(Container)
-      const mockStore = {
-        dispatch: () => {},
-        subscribe: () => {},
-        getState: () => expectedState
-      }
 
-      rtl.render(<Decorated store={mockStore} />)
+      const store1 = createStore(() => expectedState)
+      const store2 = createStore(() => ignoredState)
+
+      rtl.render(
+        <ProviderMock context={context} store={store1}>
+          <ProviderMock store={store2}>
+            <Decorated context={context} />
+          </ProviderMock>
+        </ProviderMock>
+      )
 
       expect(actualState).toEqual(expectedState)
     })
 
-    it('should throw an error if the store is not in the props or context', () => {
+    it.skip('should throw an error if the store is not in the props or context', () => {
```

Comment:
@markerikson This skip is my only concern in the 6.0.0 work. I can pick up the work if you like

Otherwise, looks great!

Reply:
I'll try to take a look at it tomorrow.  I think I disabled that test when doing some of the earlier work in #995 .  I'll need to review what the behavior is / ought to be.

Reply:
@markerikson @cellog Cool. Let me know. Seems like a simple enough piece of work

And also, the matrices could add back 16.3 and the peer-dep dropped to 16.3 like you guys mentioned

Reply:
16.3 was effectively broken with respect to `getDerivedStateFromProps`.  Anyone on 16.3 _ought_ to be able to bump to 16.4, so I don't think there's any reason to target it directly.

Reply:
Hmm.  I just pulled master, created a branch, and turned on this test... and it passes.  (Even tried changing the message to verify it fails in that case.)

I guess it was failing at some point in the process and I turned it off, but it works as intended now :)

Reply:
I'm gonna mark this as resolved, and push a change to master that just turns this test on.

Reply:
Sounds good. Glad I could help